Allow for more than 64 scheduler resources

martien-de-jong opened the pull request:
More more latest news:

QoR is done. We have about 10% of the benchmarks that gain 1 unit (16 bytes) of PM, and a very slight blip for instruction count.

I have created https://jira.xilinx.com/browse/AIECC-1021 to deal with the regression afterwards.

More Latest news:

StaticBitSet is now implemented using std::bitset.

Latest news:
This is a functionally correct version but not the definitive one. It has fixed a pre-existing bug in the hazardrecognizer that crashed due to a more strict assert. I agreed with Andreu to change ResourceSet into an interface class on top of std::bitset, with a few name changes that better align with HazardRecognizer parlance and hiding a constructor which is ambiguous and awkward in our scheme of things. 
I added unit tests for ResourceScoreboard and StaticBitSet.
There is a slight regression due to a correction of the pipelinedepth for memory operations, which tends to add nop cycles to exit blocks.

I sneak in two NFC commits that I had in a working branch. The second one centralizes all itinerary queries, which is useful for this PR. Taking the first one prevented a cherry-pick conflict.

The core of the change is that tablegen now generates the initializer for the resource through a macro, and the file that includes the generated data can change that macro.
We employ this to just list the resource index, not the expanded bitvector. This index can be converted to the ADT that can now be stretched to an arbitrary size locally in AIEHazardRecognizer.h

Commit:
[ResourceScoreboard][NFC] Slightly more generic interface.

We don't require a symmetric range around Head. This is more natural for e.g.
modulo schedules, that stretch into the future.
Instead, we configure an explicit range around Head's default position of 0.

Commit:
[AIE][NFC] Refactor AIEHazardRecognizer and Pipeliner's scoreboard usage

Commit:
[TABLEGEN][NFC] Hide resource representation in a user-definable macro

Generate a named constant to export the number of resources

Commit:
[AIE][NFC] Remove some strangeness

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -82,10 +83,12 @@ class FuncUnitWrapper {
-                                                                 : 0),
+      : Required(IS.getReservationKind() == InstrStage::Required
+                     ? (One << IS.getUnits())
+                     : 0),
```
This effectively computes the set that was generated by tablegen before.

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -99,7 +99,7 @@ void FuncUnitWrapper::clearResources() {
 }
 bool FuncUnitWrapper::isEmpty() const {
-  return Required == 0 && Reserved == 0 && Slots == 0 && MemoryBanks == 0;
+  return !Required && !Reserved && Slots == 0 && MemoryBanks == 0;
```
Not stricly necessary (anymore). 0 was interpreted as a set containing resource 0. I guess I'll add an .empty() method

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -121,9 +121,9 @@ FuncUnitWrapper &FuncUnitWrapper::operator|=(const FuncUnitWrapper &Other) {
-      (Reserved & Other.Required) != 0 || (Required & Other.Reserved) != 0 ||
+  if ((Required & Other.Required) || (Slots & Other.Slots) != 0 ||
+      (MemoryBanks & Other.MemoryBanks) != 0 || (Reserved & Other.Required) ||
+      (Required & Other.Reserved) ||
```
We may reorder these a bit, since the Required test has become more expensive.

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -57,16 +57,83 @@ const_bundled_instrs(const MachineInstr &MI, bool IncludeRoot = false) {
+  static const int Shift = 6;
+  static const int Mask = 0x3f;
+  static const uint64_t One = 1;
+  uint64_t Bits[NumContainers];
```
For initial stress testing, I'll change this temporarily to multiple uint8_t, so that aie2 and aie2p will definitely hit the second container

martien-de-jong replied:
As discussed. By building it on top of std::bitset we can delegate to std::bitset testing.

martien-de-jong commented on llvm/include/llvm/CodeGen/ResourceScoreboard.h:
```diff
@@ -70,27 +65,37 @@ template <typename RC> class ResourceScoreboard {
-    return DeltaCycles >= -Depth && DeltaCycles <= 0;
+
+  bool isInRange(int Index) const {
+    return Index >= Lowest && Index <= Highest;
```
The former applied just to the insertion point of an instruction. The latter checks each access.

martien-de-jong commented on llvm/include/llvm/CodeGen/ResourceScoreboard.h:
```diff
@@ -179,16 +179,12 @@ template <typename RC> class ResourceScoreboard {
+  // hazardrecognizer variants. In that original use, D was determined by
+  // the pipeline depth of the itineraries. We would
   // double the size and putting the origin in the middle, so that we could
   // insert in cycles < 0. isValidDelta represents this sanity check
```
nit: comment out of date.

andcarminati commented on llvm/utils/TableGen/SubtargetEmitter.cpp:
```diff
@@ -446,12 +446,17 @@ void SubtargetEmitter::EmitStageAndOperandCycleData(
 
     StringRef Name = ProcModel.ItinsDef->getName();
+    OS << "\n"
+          "#ifndef FUNCUNIT_REPRESENTATION\n"
```
Nice, this give us a clear and less intrusive way to extend.

andcarminati commented on llvm/include/llvm/CodeGen/ResourceScoreboard.h:
```diff
@@ -70,27 +65,37 @@ template <typename RC> class ResourceScoreboard {
     }
-    Depth = Pow2;
-    Size = 2 * Depth;
+    Size = Pow2;
```
Curious about the 2 * Depth that was here before. Non explainable safety margin maybe?

martien-de-jong replied:
No, that was an explicit change introduced for backward scheduling. Depth came in as ~pipeline depth, but backward scheduling wants to schedule the most urgent one first; that required a negative insertion cycle, otherwise we would need to wait until current cycle was bumped far enough. In that waiting time, shorter instructions would start eating away at the resources required to schedule the longer one.

andcarminati commented on llvm/include/llvm/CodeGen/ResourceScoreboard.h:
```diff
@@ -35,25 +35,20 @@ template <typename RC> class ResourceScoreboard {
   /// When querying for conflicts, we have more liberty, with out-of-range
   /// cycles interpreted as being empty.
-  int Depth = 0;
+  int Lowest = 0;
```
nit: maybe `LowestCycle` and `HighestCycle` to have a unit here.

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -796,4 +802,4 @@ AIEHazardRecognizer::getMemoryObjectsBits(const MachineInstr *MI) const {
   }
   return ObjectsToBitMap;
-}
+}
```
nit: new line here!

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -543,34 +568,21 @@ bool AIEHazardRecognizer::checkConflict(
-      }
+  // Note that DeltaCycles will be negative for bottom-up scheduling.
+  auto CycleConflict = [&Scoreboard,
+                        DeltaCycles](int C, const FuncUnitWrapper &ThisCycle) {
```
Can we rename this `C` to a more meaningful name?

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -543,34 +568,21 @@ bool AIEHazardRecognizer::checkConflict(
+  // Note that DeltaCycles will be negative for bottom-up scheduling.
+  auto CycleConflict = [&Scoreboard,
+                        DeltaCycles](int C, const FuncUnitWrapper &ThisCycle) {
+    int StageCycle = DeltaCycles + C;
```
nit: `const int StageCycle`...

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -632,24 +643,22 @@ void AIEHazardRecognizer::enterResources(
-  }
+  auto Insert = [&Scoreboard, DeltaCycles](int C,
+                                           const FuncUnitWrapper &ThisCycle) {
+    int ScoreboardCycle = C + DeltaCycles;
```
nit: const

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -57,16 +57,83 @@ const_bundled_instrs(const MachineInstr &MI, bool IncludeRoot = false) {
+    int get() const { return Value; }
+  };
+
+  ResourceSet() {
```
maybe just:
```
  ResourceSet() {
    for (auto &BitsPart : Bits) {
      BitsPart = 0;
    }
  }
```

martien-de-jong replied:
If that works for a low-level array, fine.

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -57,16 +57,83 @@ const_bundled_instrs(const MachineInstr &MI, bool IncludeRoot = false) {
+    return Result;
+  }
+
+  operator bool() const {
```
maybe just:
```
  operator bool() const {
    return llvm::any_of(Bits, [](auto BitsPart) { return BitsPart; });
  }
```

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -58,7 +58,9 @@ const_bundled_instrs(const MachineInstr &MI, bool IncludeRoot = false) {
 class ResourceSet {
-  static const int NumContainers = 1;
+  // This number should be big enough to accommodate
+  // all AIE targets.
```
I will look into computing the max from the tablegen files. We already collect all target itinerary inc files in a single file.

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -57,16 +57,83 @@ const_bundled_instrs(const MachineInstr &MI, bool IncludeRoot = false) {
+    return Result;
+  }
+
+  ResourceSet operator~() const {
```
Is this way too disruptive?

```
  ResourceSet operator~() const {
    ResourceSet Result;

    std::transform(std::begin(Bits), std::end(Bits), std::begin(Result.Bits),
                   [&](auto Part) { return ~Part; });

    return Result;
  }
```

martien-de-jong replied:
It's just not the way that I would like to convince the reader that this is a loop which will be unrolled to (on intel) two instructions.

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -57,16 +57,83 @@ const_bundled_instrs(const MachineInstr &MI, bool IncludeRoot = false) {
+      Bits[I] &= Other.Bits[I];
+    }
+    return *this;
+  }
```
New line here.

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -57,16 +57,83 @@ const_bundled_instrs(const MachineInstr &MI, bool IncludeRoot = false) {
+    return *this;
+  }
+
+  ResourceSet &operator&=(const ResourceSet &Other) {
```
How about this?

```
  ResourceSet &operator&=(const ResourceSet &Other) {

    std::transform(std::begin(Bits), std::end(Bits), std::begin(Other.Bits),
                   std::begin(Bits),
                   [&](auto Part, auto PartOther) { return Part & PartOther; });
    return *this;
  }
```

martien-de-jong replied:
I will not give you my honest opinion, it might hurt.

andcarminati commented:
Nice update @martien-de-jong on this infrastructure. I left some non-critical comments as a first round. Also, can we have some mock tests to stress the implementation?

Commit:
[AIE][NFC] Change the FuncUnit representation

martien-de-jong commented on llvm/utils/TableGen/SubtargetEmitter.cpp:
```diff
@@ -1507,6 +1509,32 @@ void SubtargetEmitter::EmitProcessorModels(raw_ostream &OS) {
 
+//
+// EmitNumResources
+// Emit a const expr reprenting the number of resources
```
nit: typo.

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -57,15 +58,20 @@ const_bundled_instrs(const MachineInstr &MI, bool IncludeRoot = false) {
+  return (NumResources + BitsPerContainer - 1) / BitsPerContainer;
+}
+constexpr int NumContainers = numContainers(TotalNumResources);
+} // namespace
```
AN earlier version of this commit had this in AIEMaxNumResources.h.  I think it is better to have it here.

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -57,15 +58,113 @@ const_bundled_instrs(const MachineInstr &MI, bool IncludeRoot = false) {
+  }
+
+  // Mainly for generating empty sets
+  ResourceSet(uint64_t Value) : ResourceSet() { Bits[0] = Value; }
```
Why do we have `Bits[0] = Value;` here? Can we just use `ResourceSet(uint64_t) : ResourceSet() {}`?

martien-de-jong replied:
This is a backward compatible constructor, which can initialize a 'new' bit set with an 'old' initializer  We don't really need it I guess. What do you think?

andcarminati replied:
Actually, we need this constructor, but it could be this simpler version I guess.

martien-de-jong replied:
We only needed the constructor for the empty set. I made all of those more explicit.

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -82,10 +181,12 @@ class FuncUnitWrapper {
 
   FuncUnitWrapper(const InstrStage &IS, SlotBits Slots = 0,
                   MemoryBankBits MemoryBanks = 0, uint64_t MemObjectsBits = 0)
```
Oh, this was left behind. Could you please also update `uint64_t` to `MemoryObjectsBits`?

martien-de-jong replied:
And change MemoryObjectBits to a larger set ? ;-)

andcarminati replied:
Now we have the recipe, but I think we can stick to 64bit for now ;-)

andcarminati commented on llvm/lib/Target/AIE/AIEMaxNumResources.h:
```diff
@@ -0,0 +1,31 @@
+#include "AIE2PGenSubtargetInfo.inc"
+
+constexpr const int TotalNumResources = std::max(
+    AIE2PItineraries::NumResources,
```
I like this flexibility, we don't need to guess this information. Maybe we should write a comment saying that this file should be updated for every new target.

martien-de-jong replied:
git grep -l "^#include .*\.inc" in the AIE directory?

andcarminati commented on llvm/lib/Target/AIE/AIEMaxNumResources.h:
```diff
@@ -0,0 +1,31 @@
+
+constexpr const int TotalNumResources = std::max(
+    AIE2PItineraries::NumResources,
+    std::max(AIE2Itineraries::NumResources, AIEItineraries::NumResources));
```
nit (just one max):
```
constexpr const int TotalNumResources =
    std::max({AIE2PItineraries::NumResources, AIE2Itineraries::NumResources,
              AIEItineraries::NumResources});
```

martien-de-jong replied:
Ah, that's the syntax, thanks! I was expecting some recursive forwarding construct

andcarminati commented on llvm/lib/Target/AIE/AIEMaxNumResources.h:
```diff
@@ -0,0 +1,31 @@
+#define LLVM_LIB_TARGET_AIE_AIEMAXNUMRESOURCES_H
+
+#include <algorithm>
+#include <climits>
```
Unnecessary (climits and cstdint )?

Commit:
[AIE][NFC] Header file to collect the maximum number of resources over all archs

Commit:
[AIE][NFC] Add StaticBitSet

Commit:
[AIE] Use StaticBitSet to hold resources in AIEHazardRecognizer

We simplified some constructors of FuncUnitWrapper, formalizing the
status quo that we deal with the different components separately; Only the
scoreboard collects them into cycles
We now check every access to the scoreboard rather than just the intertion
cycle. This brought to light a bug in checking the memory cycles, which
implicitly assumed that the memory cycles would correspond with some
resource being used.

martien-de-jong commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -634,24 +650,23 @@ void AIEHazardRecognizer::enterResources(
     FuncUnitWrapper MemoryBankAndObjectsAccessCycle(
         /*Req=*/0, /*Res=*/0, /*SlotSet=*/0, MemoryBanks, MemObjectsBits);
     for (auto Cycles : MemoryAccessCycles) {
+      assert(Scoreboard.isInRange(DeltaCycles + Cycles - 1));
```
nit: I guess we may want to put the assert in the `operator[]` method

martien-de-jong commented on llvm/lib/Target/AIE/StaticBitSet.h:
```diff
@@ -0,0 +1,108 @@
+  // element.
+  explicit StaticBitSet(int BitNo) : StaticBitSet() {
+    int BlockIdx = BitNo >> Shift;
+    assert(BlockIdx < NumContainers && "Adjust NumContainers");
```
nit: perhaps "Adjust NumBits"  or "BitNo out of Range"

martien-de-jong commented on llvm/lib/Target/AIE/StaticBitSet.h:
```diff
@@ -0,0 +1,108 @@
+    StaticBitSet Result(*this);
+    Result &= Other;
+    return Result;
+  }
```
I actually don't use this yet, but it looks ugly to break symmetry between & and |

martien-de-jong replied:
Also, if we start using it elsewhere, we should have a  set difference operation, i.e. a & ~b
Perhaps also an iterator, which may require masking after set negation so that we don't list entries outside of NumBits

andcarminati commented on llvm/unittests/CodeGen/ResourceScoreboardTest.cpp:
```diff
@@ -0,0 +1,143 @@
+  for (int I = 0; I <= 11; I++) {
+    EXPECT_EQ(Scoreboard[11 - I], 31 - I);
+  }
+}
```
nit: new line here.

martien-de-jong commented on llvm/unittests/CodeGen/ResourceScoreboardTest.cpp:
```diff
@@ -0,0 +1,143 @@
+//===- ResourceScoreboardTest.cpp --------------------------------*- C++
+//-*-===//
```
grmbl

andcarminati commented on llvm/unittests/Target/AIE/StaticBitSetTest.cpp:
```diff
@@ -0,0 +1,123 @@
+
+#include "StaticBitSet.h"
+
+// I like primes.
```
;-)

andcarminati commented:
As discussed offline, we can use `std::bitset` or even `llvm/ADT/Bitset.h`.

martien-de-jong commented on llvm/lib/Target/AIE/StaticBitSet.h:
```diff
@@ -0,0 +1,108 @@
+
+#include <cassert>
+#include <climits>
+
```
This is a least effort implementation of a bitset that meets our purposes. The functionality is very close to std::bitset, but we can configure the container size for stress testing, and we only have a non-default constructor that interprets its single operand as a bitindex, rather than as a bitset. This meshes better with the resource interpretation as we use it.

Commit:
[AIE][NFC] Rewrite StaticBitSet on top of std::bitset

andcarminati commented on llvm/lib/Target/AIE/StaticBitSet.h:
```diff
@@ -0,0 +1,55 @@
+
+#include <bitset>
+
+template <const int NumBits> class StaticBitSet : public std::bitset<NumBits> {
```
Sounds good!

andcarminati commented:
> There is a slight regression due to a correction of the pipelinedepth for memory operations, which tends to add nop cycles to exit blocks.

Hi @martien-de-jong, does it mean that we were optimistic before? I am curious about the QoR effect.

andcarminati commented on llvm/lib/Target/AIE/StaticBitSet.h:
```diff
@@ -0,0 +1,55 @@
+// (c) Copyright 2025 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+// This is a very low level implementation of a bitset, designed to be fully
```
I think this is in a higher level now with `std::bitset`.

martien-de-jong replied:
true. I'll adjust.

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -85,28 +85,28 @@ void FuncUnitWrapper::dump() const {
 bool FuncUnitWrapper::isEmpty() const {
-  return Required == 0 && Reserved == 0 && Slots == 0 && MemoryBanks == 0 &&
-         MemObjectsBits == 0;
+  return Required.empty() && Reserved.empty() && Slots == 0 &&
```
I expect this `.empty()` to be fast, but not faster than `== 0`. Should we change the order here?

martien-de-jong replied:
I find zero references.... I'll try to remove it.

martien-de-jong replied:
Oh, differently unimportant. Only used from dumpers.

andcarminati commented on llvm/unittests/CodeGen/ResourceScoreboardTest.cpp:
```diff
@@ -0,0 +1,142 @@
+// (c) Copyright 2025 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
```
Perhaps a description of the goal of this test, for the future!

F-Stuckmann commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -623,35 +640,33 @@ void AIEHazardRecognizer::enterResources(
-    Cycle += IS.getNextCycles();
-  }
+  // Insert ThisCycle at position Cycle relative to the start of the itinerary.
+  auto Insert = [&Scoreboard, DeltaCycles](int Cycle,
```
Is this the same as reserving a FuncUnit? Could we harmonize the names?

martien-de-jong replied:
Discussed offline. It's inserting a full FuncUnitWrapper into the scoreboard. This represents one cycle of the itinerary, in this case the Resaerved and Required resources.

F-Stuckmann commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -671,14 +686,13 @@ void AIEHazardRecognizer::computeMaxLatency() {
   unsigned FirstRW = std::numeric_limits<unsigned>().max();
   unsigned LastRW = 0;
-  unsigned ItinDepth = 0;
+  auto MaxDepth = [this](int C, const FuncUnitWrapper &) {
```
we only access PipelineDepth , could we only Capture the reference PipelineDepth ?

martien-de-jong replied:
But I can't capture class members....

andcarminati commented on llvm/lib/Target/AIE/AIEPostPipeliner.h:
```diff
@@ -210,9 +210,8 @@ class PostPipeliner {
   int RecMII = 0;
 
-  // The scoreboard and its depth.
+  // The scoreboard
```
nit: `The scoreboard.`

andcarminati commented on llvm/test/CodeGen/AIE/aie2/schedule/postpipeliner/add-store.mir:
```diff
@@ -40,8 +40,10 @@
   ; CHECK-NEXT:  // %bb.3: // %for.cond.cleanup
-  ; CHECK-NEXT:    nopb ; nopa ; st r1, [p0], #4; nopxm ; nopv
-  ; CHECK-NEXT:    nopx
+  ; CHECK-NEXT:    nopa ; nopb ; nopxm ; st r1, [p0], #4
```
I guess that 9 cycles here is too much!

andcarminati replied:
Maybe not ;-)

andcarminati commented on llvm/unittests/Target/AIE/HazardRecognizerTest.cpp:
```diff
@@ -24,24 +24,48 @@ using namespace llvm;
 const auto Res = InstrStage::ReservationKinds::Reserved;
 
+// Note: This test has a history, and bears the scratches and tears of
+// a number of hazardrecognizer and scoreboard reorganizations
```
nit: `reorganizations.`

F-Stuckmann commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -137,6 +138,31 @@ bool FuncUnitWrapper::conflict(const FuncUnitWrapper &Other) const {
+// returning true as soon as Action returns true.
+// A full traversal can be made by returning false.
+bool anyStage(ArrayRef<const InstrStage> Stages,
+              std::function<bool(int C, const FuncUnitWrapper &)> Action,
```
If you define the callable, you can use them in the lambdas to clarify them. 

```
using FuncUnitAction = std::function<bool(int C, const FuncUnitWrapper &)>;
bool anyStage(ArrayRef<const InstrStage> Stages, FuncUnitAction Action,
              std::optional<int> FUDepthLimit = std::nullopt) {
```
Then in the lambdas:
```
  FuncUnitAction Insert = [&Scoreboard, DeltaCycles](int Cycle,
                                           const FuncUnitWrapper &ThisCycle) {
    const int ScoreboardCycle = Cycle + DeltaCycles;
    assert(Scoreboard.isInRange(ScoreboardCycle));
    Scoreboard[ScoreboardCycle] |= ThisCycle;
    return false;
  };
```

martien-de-jong replied:
Thanks for trying that!

F-Stuckmann commented on llvm/unittests/CodeGen/CMakeLists.txt:
```diff
@@ -1,3 +1,10 @@
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# (c) Copyright 2025 Advanced Micro Devices, Inc. or its affiliates
```
NIT: Copyright should be 
// Modifications (c) Copyright 2025 Advanced Micro Devices, Inc. or its
// affiliates

martien-de-jong commented:
> > There is a slight regression due to a correction of the pipelinedepth for memory operations, which tends to add nop cycles to exit blocks.
> 
> Hi @martien-de-jong, does it mean that we were optimistic before? I am curious about the QoR effect.

As discussed, we were wrong, if we want to use a 'pure' definition of the pipeline depth. The memory cycles were recorded in the scoreboard, and the allocation of the scoreboard uses the pipelinedepth. Since we didn't take the memory cycles into account for the pipelinedepth, we were checking an unrelated cycle for the resource conflict, which can lead to both false negatives and false positives. The pipelinedepth is also used to define the safety margin when we jump into an unknown block, hence these new extra cycles. The latter use of the pipelinedepth  was safe, since there we implicitly used the fact that all instructions with memory cycles do that late in the pipeline, ~E4.
I think we need two notions of the pipelinedepth, one that defines the range the scoreboard should be able to access. the other to represent the safely margin.

andcarminati commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -709,6 +711,7 @@ void AIEHazardRecognizer::computeMaxLatency() {
   FirstRW = TII->getMinFirstMemoryCycle();
   LastRW = TII->getMaxLastMemoryCycle();
   MaxLatency = std::max(MaxLatency, int(LastRW - FirstRW + 1));
+  PipelineDepth = std::max(PipelineDepth, int(LastRW));
```
Oh, here is the bug fix that was mentioned.

andcarminati reviewed (approved):
LGTM. Nice work!
